hardware/awg/tektronix_awg70k.py

```diff
@@ -1678,3 +1723,487 @@ def _create_xml_header(self, number_of_samples, markers_active):
+                return False
+            return True
+        return False
+
+    def _get_mw_trigger_str(self):
+        if self._microwave_trigger in self.__event_triggers:
+            trigger = self.__event_triggers[self._microwave_trigger]
+        elif self._microwave_trigger in self.__event_triggers.values():
+            trigger = self._microwave_trigger
+        else:
+            raise Exception('Invalid "microwave_trigger" found. Unable to adjust trigger polarity.')
+        return trigger
+
+    def _init_waveform_compiler(self):
+        if 'Basic Waveform' not in self.query('WPL:PLUG?'):
+            raise Exception('Waveform creation plug-in "Basic Waveform" not installed on device.')
+        self.write('WPL:ACT "Basic Waveform"')
+        self.write('*WAI')
+        self.query('*OPC?')
+        self.write('BWAV:FUNC SINE')
+        self.write('*WAI')
+        self.query('*OPC?')
+        self.write('BWAV:AUTO LEN')
+        self.write('*WAI')
+        self.query('*OPC?')
+        self.write('BWAV:COMP:CASS 0')
+        self.write('*WAI')
+        self.query('*OPC?')
+        self.write('BWAV:COMP:PLAY 0')
+        self.write('*WAI')
+        self.query('*OPC?')
+        self.write('BWAV:FDR 0')
+        self.write('*WAI')
+        self.query('*OPC?')
+        self.write('BWAV:OFFS 0')
+        self.write('*WAI')
+        self.query('*OPC?')
+        return
+
+    def _compile_sine_waveform(self, name, frequency, amplitude, cycles):
+        frequency = float(frequency)
+        amplitude = float(amplitude)
+        cycles = int(cycles)
+        # Set values
+        self.write('BWAV:COMP:NAME "{0}"'.format(name))
+        self.query('*OPC?')
+        self.write('BWAV:AMPL {0:.9e}'.format(amplitude * 2))
+        self.query('*OPC?')
+        self.write('BWAV:FREQ {0:.9e}'.format(frequency))
+        self.query('*OPC?')
+        self.write('BWAV:CYCL {0:d}'.format(cycles))
+        self.query('*OPC?')
+
+        # Read back values
+        set_name = self.query('BWAV:COMP:NAME?')
+        set_freq = float(self.query('BWAV:FREQ?'))
+        set_amp = float(self.query('BWAV:AMPL?')) / 2
+        set_cycles = int(float(self.query('BWAV:CYCL?')))
+
+        # Try to set the values again if there has been an error. This is Tektronix bullsh..
+        if set_freq != frequency or set_cycles != cycles:
+            print(self.read_error_register())
+            self.write('BWAV:FREQ {0:.9e}'.format(frequency))
+            self.query('*OPC?')
+            self.write('BWAV:CYCL {0:d}'.format(cycles))
+            self.query('*OPC?')
+            set_freq = float(self.query('BWAV:FREQ?'))
+            set_cycles = int(float(self.query('BWAV:CYCL?')))
+
+        # Compile waveform and set laser channel marker to high
+        self.write('BWAV:COMP')
+        self.write('*WAI')
+        self.query('*OPC?')
+        a_ch = int(name.rsplit('_ch', 1)[-1])
+        laser_ch = int(self._laser_channel.rsplit('ch', 1)[-1])
```

Comment:
What happens if the laser channel is empty?

Reply:
Well first of all it's a protected variable so any changes you do to it from outside the module are generally at your own risk.

I added some sanity checks to the activation to avoid corrupted AppStatus.